File: app/lib/features/calendar/models/calendar_event.dart
class CalendarEvent {
  final String id;
  final String householdId;
  final String title;
  final String? description;
  final String eventType;
  final DateTime startTime;
  final DateTime? endTime;
  final bool allDay;
  final String? recurrenceRule;
  final String? location;
  final double? locationLat;
  final double? locationLng;
  final String? sourceId;
  final String? sourceType;
  final String createdBy;
  final DateTime createdAt;

  CalendarEvent({
    required this.id,
    required this.householdId,
    required this.title,
    this.description,
    required this.eventType,
    required this.startTime,
    this.endTime,
    required this.allDay,
    this.recurrenceRule,
    this.location,
    this.locationLat,
    this.locationLng,
    this.sourceId,
    this.sourceType,
    required this.createdBy,
    required this.createdAt,
  });

  factory CalendarEvent.fromJson(Map<String, dynamic> json) {
    return CalendarEvent(
      id: json['id'] as String,
      householdId: json['household_id'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      eventType: json['event_type'] as String? ?? 'other',
      startTime: DateTime.parse(json['start_time'] as String),
      endTime: json['end_time'] != null
          ? DateTime.parse(json['end_time'] as String)
          : null,
      allDay: json['all_day'] as bool? ?? false,
      recurrenceRule: json['recurrence_rule'] as String?,
      location: json['location'] as String?,
      locationLat: (json['location_lat'] as num?)?.toDouble(),
      locationLng: (json['location_lng'] as num?)?.toDouble(),
      sourceId: json['source_id'] as String?,
      sourceType: json['source_type'] as String?,
      createdBy: json['created_by'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'event_type': eventType,
      'start_time': startTime.toIso8601String(),
      'end_time': endTime?.toIso8601String(),
      'all_day': allDay,
      'recurrence_rule': recurrenceRule,
      'location': location,
      'location_lat': locationLat,
      'location_lng': locationLng,
    };
  }

  CalendarEvent copyWith({
    String? title,
    String? description,
    String? eventType,
    DateTime? startTime,
    DateTime? endTime,
    bool? allDay,
    String? recurrenceRule,
    String? location,
    double? locationLat,
    double? locationLng,
  }) {
    return CalendarEvent(
      id: id,
      householdId: householdId,
      title: title ?? this.title,
      description: description ?? this.description,
      eventType: eventType ?? this.eventType,
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
      allDay: allDay ?? this.allDay,
      recurrenceRule: recurrenceRule ?? this.recurrenceRule,
      location: location ?? this.location,
      locationLat: locationLat ?? this.locationLat,
      locationLng: locationLng ?? this.locationLng,
      sourceId: sourceId,
      sourceType: sourceType,
      createdBy: createdBy,
      createdAt: createdAt,
    );
  }
}

File: backend/app/models/calendar_event.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import Boolean, DateTime, Float, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from ..core.database import Base


class CalendarEvent(Base):
    __tablename__ = "calendar_events"

    id: Mapped[str] = mapped_column(UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4()))
    household_id: Mapped[str] = mapped_column(UUID(as_uuid=False), nullable=False, index=True)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    event_type: Mapped[str] = mapped_column(String(20), default="other", nullable=False)
    start_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    end_time: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    all_day: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    recurrence_rule: Mapped[str | None] = mapped_column(String(100), nullable=True)
    location: Mapped[str | None] = mapped_column(String(500), nullable=True)
    location_lat: Mapped[float | None] = mapped_column(Float, nullable=True)
    location_lng: Mapped[float | None] = mapped_column(Float, nullable=True)
    source_id: Mapped[str | None] = mapped_column(UUID(as_uuid=False), nullable=True)
    source_type: Mapped[str | None] = mapped_column(String(20), nullable=True)
    created_by: Mapped[str] = mapped_column(UUID(as_uuid=False), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
